swarm/pss/pss.go

```diff
@@ -241,91 +317,283 @@ func (self *Pss) checkFwdCache(addr []byte, digest pssDigest) bool {
+}
+
+// Sends a pss message
+//
+// The method itself is payload agnostic, and will accept any arbitrary byte slice as the payload for a message.//
+// It generates an whisper envelope for the specified recipient and topic, and wraps the message payload in it.
+func (self *Pss) send(to []byte, topic whisper.TopicType, msg []byte, pubkey *ecdsa.PublicKey, symkey []byte) error {
+	if pubkey != nil && symkey != nil {
+		return fmt.Errorf("Can only specify one of symkey and pubkey for send")
+	} else if pubkey == nil && symkey == nil {
+		return fmt.Errorf("No keys set, refusing to send unencrypted pss")
+	}
+	wparams := &whisper.MessageParams{
+		TTL:      DefaultTTL,
+		Src:      self.privatekey,
+		Dst:      pubkey, // if this is set, the message is considered asymmetric, if not it is considered symmetric
+		KeySym:   symkey,
+		Topic:    topic,
+		WorkTime: defaultWhisperWorkTime,
+		PoW:      defaultWhisperPoW,
+		Payload:  msg,
+		Padding:  []byte("1234567890abcdef"),
+	}
+	// set up outgoing message container, which does encryption and envelope wrapping
+	woutmsg, err := whisper.NewSentMessage(wparams)
+	if err != nil {
+		return fmt.Errorf("failed to generate whisper message encapsulation: %v", err)
+	}
+	// performs encryption and PoW
+	// after this the message is ready for sending
+	envelope, err := woutmsg.Wrap(wparams)
+	if err != nil {
+		return fmt.Errorf("failed to perform whisper encryption: %v", err)
+	}
+	log.Trace("pssmsg whisper done", "env", envelope, "wparams payloda", wparams.Payload)
+	// prepare for devp2p transport
 	pssmsg := &PssMsg{
 		To:      to,
-		Payload: pssenv,
+		Payload: envelope,
 	}
 	return self.Forward(pssmsg)
 }
 
-// Forwards a pss message to the peer(s) closest to the to address
+// Forwards a pss message to the peer(s) closest to the to recipient address in the PssMsg struct
+//
+// The To-field in the PssMsg struct will be plaintext, unencrypted, and will be truncated to the magnitude set by SetRecipientAddressLength (PssParams.RecipientAddressLength by default).
+// If the recipientAddressLength value is 0, no bytes will be disclosed, and message will be forwarded to ALL devp2p peers
 //
-// Handlers that are merely passing on the PssMsg to its final recipient should call this directly
+// Handlers that are merely passing on the PssMsg to its final recipient might call this directly
 func (self *Pss) Forward(msg *PssMsg) error {
-
-	if self.isSelfRecipient(msg) {
-		return errorForwardToSelf
-	}
-
-	// cache it
+	to := msg.To
+	zeros := make([]byte, len(to)-int(self.recipientAddressLength))
+	// truncate part of the pssmsg recipient address
+	// this way noone can know for certain exactly who it's for
+	// the kademlia address will be same length
+	// but equivalent to truncated part will be overwritten by zeros
+	copy(to[self.recipientAddressLength:len(to)], zeros)
+	log.Trace("truncated recipient", "original", msg.To, "truncated", to)
+	msg.To = msg.To[:self.recipientAddressLength]
```

Comment:
I don't quite understand what you mean by "check" - the logic just truncates the address. But the kademlia iterator doesn't allow for addresses < 32 bytes (at least i got panics when trying).